[optim]fix ut and sgd kernel

zhuhaozhe opened the pull request:
- Original `test_grad_scaling_autocast_fused_optimizers` does not work since there is no "fused" in `optim_inputs`
 - We should use different `grad_scaler`, they should not share 1 `scale`, there is no issue exposed here because the default `_growth_interval` is 2000 so it will not growth and there is also no inf is found so it will not reduced. The one in `test_cuda.py` should also have this issue,
 - I set a manual seed to reproduce purpose if there is any numerical failure
 - I use Tensor tracker here because we failed this UT in dynamo case, the cpp generated code are not exactly same with fused/non fused kernel.
 - I make it check both `cuda` and `cpu`.
 - I find some SGD numerical issue with `clang`, and fixed it by using `fmadd` instead of `add/mul` in fused sgd veckernel.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124904



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

zhuhaozhe committed:
[optim]fix ut and sgd kernel

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[optim]fix ut and sgd kernel"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[optim]fix ut and sgd kernel"

 - Original `test_grad_scaling_autocast_fused_optimizers` does not work since there is no "fused" in `optim_inputs`
 - We should use different 




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[optim]fix ut and sgd kernel"

 - Original `test_grad_scaling_autocast_fused_optimizers` does not work since there is no "fused" in `optim_inputs`
 - We should use different 




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

janeyx99 commented on test/test_optim.py:
```diff
@@ -1704,38 +1706,44 @@ def _grad_scaling_autocast_fused_optimizers(self, optimizer_ctor, optimizer_kwar
-                scaler.unscale_(opt_scaling)
-            scaler.step(opt_scaling)
-            scaler.update()
+                scaler_scaling.unscale_(opt_scaling)
```
yikes, good catch! does this mean the original test in test_cuda is wrong today then? 

could we unify the test now?

janeyx99 replied:
Yes, we should eventually move the test back to test_cuda.py under TestCudaOptim to keep gradscaling together but let's keep in test_optim.py for now as there's a lot of movement in that file recently :D

janeyx99 commented on aten/src/ATen/native/cpu/FusedSGDKernel.cpp:
```diff
@@ -142,20 +141,20 @@ typename std::enable_if<
-            Vec::loadu(momentum_buf_ptr + d) * Vec(scalar_t(momentum)) +
-            grad_vec * Vec(scalar_t(1 - dampening));
+            Vec::loadu(momentum_buf_ptr + d) * Vec(scalar_t(momentum));
+        momentum_vec = vec::fmadd(Vec(scalar_t(1 - dampening)), grad_vec, momentum_vec);
```
for my understanding: what’s the difference between vec::fmadd vs += and a *? I’m guessing it’s for performance (this is a fused kernel) but does it also do more vectorization?

zhuhaozhe replied:
I think there might be some rounding differences between `fma` and  `+ and *`.

janeyx99 reviewed (approved):
approving to rectify the testing situation asap, but I would like to see us unify the test between test_cuda and test_optim sooner rather than later.

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe committed:
Update on "[optim]fix ut and sgd kernel"

 - Original `test_grad_scaling_autocast_fused_optimizers` does not work since there is no "fused" in `optim_inputs`
 - We should use different `grad_scaler`, they should not share 1 `scale`, there is no issue exposed here because the default `_growth_interval` is 2000 so it will not growth and there is also no inf is found so it will not reduced. The one in `test_cuda.py` should also have this issue,
 - I set a manual seed to reproduce purpose if there is any numerical failure
 - I use Tensor tracker here because we failed this UT in dynamo case, the cpp generated code are not exactly same with fused/non fused kernel.
 - I make it check both `cuda` and `cpu`.
 - I find some SGD numerical issue with `clang`, and fixed it by using `fmadd` instead of `add/mul` in fused sgd veckernel.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge
